src/nnet3/nnet-utils.cc

```diff
@@ -26,25 +26,142 @@ namespace nnet3 {
 
 int32 NumOutputNodes(const Nnet &nnet) {
   int32 ans = 0;
-  for (int32 n = 0; n < nnet.NumNodes(); n++)
+  for (int32 n = 0; n < nnet.NumNodes(); n++) {
     if (nnet.IsOutputNode(n))
       ans++;
+  }
   return ans;
 }
 
 int32 NumInputNodes(const Nnet &nnet) {
   int32 ans = 0;
-  for (int32 n = 0; n < nnet.NumNodes(); n++)
+  for (int32 n = 0; n < nnet.NumNodes(); n++) {
     if (nnet.IsInputNode(n))
       ans++;
+  }
   return ans;
 }
 
+void GetRecurrentOutputNodeNames(const Nnet &nnet,
+                                 std::vector<std::string>
+                                 *recurrent_output_names,
+                                 std::vector<std::string>
+                                 *recurrent_node_names) {
+  recurrent_output_names->clear();
+  recurrent_node_names->clear();
+  for (int32 i = 0; i < nnet.NumNodes(); i++) {
+    if (nnet.IsOutputNode(i) && nnet.GetNodeName(i) != "output") {
+      recurrent_output_names->push_back(nnet.GetNodeName(i));
+
+      std::vector<int32> node_indexes;
+      nnet.GetNode(i).descriptor.GetNodeDependencies(&node_indexes);
+      KALDI_ASSERT(node_indexes.size() == 1);
+      recurrent_node_names->push_back(nnet.GetNodeName(node_indexes[0]));
+    }
+  }
+}
+
+void GetRecurrentNodeOffsets(const Nnet &nnet,
+                             const std::vector<std::string>
+                             &recurrent_node_names,
+                             std::vector<int32> *recurrent_offsets) {
+  recurrent_offsets->clear();
+  recurrent_offsets->resize(recurrent_node_names.size(), 0);
+  for (int32 i = 0; i < nnet.NumNodes(); i++) {
+    const NetworkNode &node = nnet.GetNode(i);
+    // skip output node to circumvent the problem when a recurrent output 
+    // is "label delayed" and thus also contain keyword "Offset", which is not
+    // what we are trying to look at for our "offset"  
+    if (node.node_type == kDescriptor && !nnet.IsOutputNode(i)) {
+      for (int32 p = 0; p < node.descriptor.NumParts(); p++) {
+        const SumDescriptor &this_part = node.descriptor.Part(p);
+        GetIntoSumDescriptor(nnet, this_part, recurrent_node_names,
+                             recurrent_offsets);
+      }
+    }
+  }
+}
+
+void GetIntoSumDescriptor(const Nnet &nnet,
```

Comment:
Choose a better name for these GetInto* functions. It is not clear from the name what these are doing. Remember to add a comment above each function describing what it is supposed to be doing.